packages/astro/src/transitions/router.ts

```diff
@@ -0,0 +1,394 @@
+		const newAnimations = document
+			.getAnimations()
+			.filter((a) => !currentAnimations.includes(a) && !isInfinite(a));
+		const finished = Promise.all(newAnimations.map((a) => a.finished));
+		const fallbackSwap = () => {
+			swap();
+			document.documentElement.dataset.astroTransitionFallback = 'new';
+		};
+		await finished;
+		fallbackSwap();
+	} else {
+		swap();
+	}
+}
+
+async function transition(dir: Direction, loc: URL, state?: State) {
+	let finished: Promise<void>;
+	const href = loc.href;
+	const { html, ok, mediaType, redirected } = await getHTML(href);
+	// if there was a redirection, show the final URL in the browser's address bar
+	redirected && (loc = new URL(redirected));
+	// If there is a problem fetching the new page, just do an MPA navigation to it.
+	if (!ok || !(mediaType === 'text/html' || mediaType === 'application/xhtml+xml')) {
+		location.href = href;
+		return;
+	}
+
+	const doc = parser.parseFromString(html, mediaType);
+	if (!doc.querySelector('[name="astro-view-transitions-enabled"]')) {
+		location.href = href;
+		return;
+	}
+
+	// Now we are sure that we will push state, and it is time to create a state if it is still missing.
+	!state && history.replaceState({ index: currentHistoryIndex, scrollX, scrollY }, '');
+
+	document.documentElement.dataset.astroTransition = dir;
+	if (supportsViewTransitions) {
+		finished = document.startViewTransition(() => updateDOM(doc, loc, state)).finished;
+	} else {
+		finished = updateDOM(doc, loc, state, getFallback());
+	}
+	try {
+		await finished;
+	} finally {
+		// skip this for the moment as it tends to stop fallback animations
+		// document.documentElement.removeAttribute('data-astro-transition');
+		await runScripts();
+		markScriptsExec();
+		onPageLoad();
+	}
+}
+
+export function navigate(href: string) {
+	const link = new URL(href, location.href);
+	// We do not need to handle same page links because there are no page transitions
+	// Same page means same path and same query params (but different hash)
+	if (location.pathname === link.pathname && location.search === link.search) {
+		if (link.hash) {
+			location.href = link.href;
+		} else {
+			// Special case: self link without hash
+			// If handed to the browser it will reload the page
+			// But we want to handle it like any other same page navigation
+			// So we scroll to the top of the page but do not start page transitions
+			// push state on the first navigation but not if we were here already
+			if (location.hash) {
+				history.replaceState({ index: currentHistoryIndex, scrollX, scrollY: -(scrollY + 1) }, '');
+				const newState: State = { index: ++currentHistoryIndex, scrollX: 0, scrollY: 0 };
+				history.pushState(newState, '', link.href);
+			}
+			scrollTo({ left: 0, top: 0, behavior: 'instant' });
+		}
+		return;
+	}
+	// these are the cases we will handle: same origin, different page
+	transition('forward', link);
+}
+
+addEventListener('popstate', (ev) => {
```

Comment:
Curious why this listener and everything below is in this file but the click listener is not. Should it all be in here? I could see leaving prefetch out, as that's not a router behavior.

Reply:
My intention was to move everything into router.ts that is necessary to use the API. For example, the popstate handler is essential because it consumes the history state set by `navigate()`. The click handler is not essential for the router. Rather the other way around. However, this scheme breaks with the meta elements. In the end, `router.ts` also depends on ViewTransition.astro, as it needs to know which pages will/could support view transitions.